app/Platform/Services/UserAgentService.php

```diff
@@ -0,0 +1,183 @@
+                'clientVersion' => $this->trimVersion(substr($clause, $index + 1)),
+            ];
+        }
+
+        // assuming first word is client and last word is version
+        $parts = explode(' ', $clause);
+        if (count($parts) > 1) {
+            return [
+                'client' => $parts[0],
+                'clientVersion' => $this->trimVersion($parts[count($parts) - 1]),
+            ];
+        }
+
+        // special case: 'libretro' => 'RetroArch'
+        if ($clause === 'libretro') {
+            return [
+                'client' => 'RetroArch',
+                'clientVersion' => 'Unknown',
+            ];
+        }
+
+        // did not find a version number at the end of the string, just return
+        // the whole clause as the client
+        return [
+            'client' => $clause ? $clause : 'Unknown',
+            'clientVersion' => 'Unknown',
+        ];
+    }
+
+    private function addSecondaryInformation(array &$data, string $clause): void
+    {
+        $index = strpos($clause, '/');
+        if ($index !== false) {
+            // found "Client/Version", just split it
+            $thing = trim(substr($clause, 0, $index));
+            $version = trim(substr($clause, $index + 1));
+
+            if ($thing === 'Integration') {
+                $data['integrationVersion'] = $version;
+            } else {
+                if (!array_key_exists('extra', $data)) {
+                    $data['extra'] = [];
+                }
+
+                $data['extra'][$thing] = $this->trimVersion($version);
+            }
+        }
+
+        $potentialVersion = $this->trimVersion($clause);
+        if ($this->looksLikeVersion($potentialVersion) && !$this->looksLikeVersion($data['clientVersion'])) {
+            $data['clientVersion'] = $potentialVersion;
+        }
+    }
+
+    private function trimVersion(string $version): string
+    {
+        $version = trim($version);
+        $version = ltrim($version, 'vV');
+
+        return $version;
+    }
+
+    private function looksLikeVersion(string $version): bool
+    {
+        // look for string starting with "N.N"
+        $len = strlen($version);
+        if ($len >= 3) {
+            // match leading digits
+            $index = 0;
+            do {
+                $c = substr($version, $index, 1);
+                if (!is_numeric($c)) {
+                    break;
+                }
+                $index++;
+            } while ($index < $len);
+
+            if ($index > 0 && $index < $len - 1) {
+                // match decimal
+                $c = substr($version, $index, 1);
+                if ($c === '.') {
+                    // match trailing digits
+                    $c = substr($version, $index + 1, 1);
+                    if (is_numeric($c)) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
```

Comment:
I believe this entire function can be implemented in 1 line with a regular expression:

```php
return preg_match('/^\d+(?:\.\d+)+/', $version) === 1;
```

Your test suite is quite comprehensive, and they all still seem to pass.

Reply:
done, but I've kept the minimum length and `is_numeric` check on the first character to avoid compiling the regex when not necessary (most strings passed to this function won't start with a number).